walkthrough/step05.md

```diff
@@ -0,0 +1,190 @@
+
+&nbsp;
+
+<img src="https://miro.medium.com/max/1142/1*fbe04fcynkBuLo_CADxxHQ.png" alt="" />
+
+### An example of a application level middleware
+
+```javascript=
+
+// middleware logging timestamp before handling requests
+app.use((req, res, next) => {
+  console.log(Date.now(), 'before');
+  next();
+});
+
+// handle '/' get requests with 'hello world'
+// and pass control to the next middleware
+app.get('/', (req, res, next) => {
+  res.send('Hello world');
+  next();
+});
+
+// middleware logging timestamp after handling requests
+app.use((req, res) => {
+  console.log(Date.now(), 'after');
+});
+
+```
+
+since express runs synchronously it will always run 
+```javascript=
+app.use((req, res, next) => {
+  console.log(Date.now(), 'before');
+  next();
+});
+
+```
+
+first and the `next()` will call the next middleware, which in this case is the `app.get`
+
+Now let's see a different example
+
+### An example of route specific middleware
+
+```javascript=
+
+const validateNewUser = (req ,res, next) => {
+    const {username, password, confirmPassword } = req.body
+    
+    // if username and password are empty end the request    
+    if (!password && !username) {
+        return res.send('Please insert a username and a password')
+    }
+
+    //  if password and confirmPassword are not identical end the request   
+    if (password !== confirmPassword) {
+        return res.send('Passwords do not match')
+    }
+    
+    // else continue to the next middleware 
+    next()
+    
+}
+
+app.post('/create-user', validateNewUser, (req, res, next) => {
+
+    // Add new user to our database
+    // and end request    
+});
+
+```
+
+In this example the `validateNewUser` middleware is the second argument for `app.post`, you can have as many functions or middlewares as you want. for example:
+
+```javascript=
+
+app.post('/create-user', validateNewUser, anotherMiddleware, andAnother, (req, res, next) => {
+
+    // Add new user to our database
+    // and end request    
+});
+
+```
+
+Or you can have them in an array:
+ 
+```javascript=
+app.post('/create-user', [validateNewUser, anotherMiddleware, andAnother], (req, res, next) => {
+
+    // Add new user to our database
+    // and end request    
+});
+
+```
+
+**important**: all of the middlewares above need to either run the `next` method or end the request. so your server does not hang.
+
+Middlewares are very powerful when your servers get really big and you don't want to repeat yourself, it is totally valid and even encouraged to have a folder just for your middlewares.
+
+# Tasks:
+
+#### 1. build the auth middleware
+
+Let's use what we learned about middlewares above to implement an auth middleware that checks if there is actual cookie and if it's valid.
+
+- we have a middlewares folder with an index file, create an `authCheck` file that exports a function.
+- in the `authCheck` function take the code from `home.js` that checks for the cookie and validates it.
```

Comment:
authCheck is both a file and the function exported?\

Reply:
well the file only exports a function, so technically both the function and file are `authCheck` but i can see how this would be confusing, so i'll change